### PR TITLE
[PLAT-6041] Create a command to add demo references to Emunittest

### PR DESCRIPTION
**Pull request by rachelmsimm**
What:
- Adds a command to add demos to Emunittest
- The demos are stored in a local JSON file
- The add command will check whether there are demos in the demo_list.json that already have the same name or key as the demo that the user is trying to add and reject it if it does

Purpose (Why):
- Adding demos with the same name or key as demos that are already in the list breaks the `autorun` functionality of Emunittest - basically, the demo won't autorun when the autorun argument is passed through as a query parameter/command line argument if the demo(s) selected to run shares a name or key with another demo in the list
- Manually adding to the demo list in the index.js file is tedious and error prone
- This allows us to add demos in a more reliable way rather than needing to manually edit the javascript file each time we want to inform Emunittest of a new demo

Testing:
- The logic is pretty simple, but I've added unit tests under `add_demo_src/tests` for the successful case and the case where a user attempts to add a demo with a duplicate name or key. They use the PyTest framework and are passing:

1. test_add_demo_success
Expected: insert_demo was called once
Results: insert_demo was called once

2. test_add_demo_key_and_name_are_valid
Expected: key_and_name_are_valid returns False
Results: key_and_name_are_valid returns False

- Manually tested the above cases and confirmed that the demo_list.json file looks as we would expect and demos added by the command are displayed as we would expect on the index.html page
- This was tested on my Mac M1 laptop with Emunittest running on Chrome

What it is not:
- This specific PR does not remove the need to add references to the Emunittest specific files to a build (emtimer.js, cpuprofiler.js, etc)
- It doesn't move the actual demo into the demos folder - it just adds a reference to it in the json file that index.js pulls from so that Emunittest knows about it and displays it on the index page. I think we would probably want to add that functionality eventually
- It doesn't validate whether the url path we give it is a valid path in your file system. I think it makes sense to implement with the above bullet

This PR adds functionality that is intended to fix [this bug](https://jira.unity3d.com/browse/PLAT-6528) after we integrate it into WPMF, and also accomplishes [this ticket](https://jira.unity3d.com/browse/PLAT-6041).

### CODE REVIEW

**Comment by rachelmsimm** [comment]
Once populated, the demo_list.json looks something like this:
Screenshot 2023-11-08 at 10 47 30 AM

File: .gitignore
Comment (juj) [inline .gitignore:7]: The line `__pycache__` will take care of these lines 6 and 7, there is no need to specify `subdir/__pycache__` again.

File: .gitignore
Comment (juj) [inline .gitignore:10]: This PR adds a file `demo_list.json`, but then places this file in .gitignore? I wonder if this is done due to the mocked unit test? If so, maybe the whole test working directory where modifications are done could instead be added to .gitignore (rather than singling out individual test files that are modified)?

File: add_demo_src/add_demo.py
Comment (juj) [reply add_demo_src/add_demo.py:1]: Or rather, remove the file `exceptions.py` completely? It doesn't seem to be used except for checking that one duplication in adding a demo?

File: add_demo_src/add_demo.py
Comment (juj) [inline add_demo_src/add_demo.py:18]: This might be shorter as
```py
def get_existing_demos():
  return json.load(open('../demo_list.json'))
```

File: add_demo_src/add_demo.py
Comment (juj) [inline add_demo_src/add_demo.py:26]: This gets me to think that there is maybe no sense in having both demo names and keys at all. There is nothing that emunittest uses these for any internal/special purpose - originally demo key was to distinguish duplicates, and demo name was the user-visible representation.

But it would be simpler to just have demo key be equal to demo name (which, effectively, means not having a separate demo key at all).

File: add_demo_src/demo.py
Comment (juj) [inline add_demo_src/demo.py:26]: I wonder if we need a Demo class at all? We could just use a regular dictionary object to represent a Demo?

File: add_demo_src/tests/test_add_demo.py
Comment (juj) [inline add_demo_src/tests/test_add_demo.py:31]: Nice to see a test, although this seems quite overkill to test a single dictionary addition. If there is a plan to add more tests, that's fine, although if it makes faster/simpler, then just nuke the test.

**Comment by juj** [comment]
I see the base of this PR is to land on top of Andrew Bianchi's feature branch: [abianchi/PLAT-5697](https://github.com/Unity-Technologies/emunittest/tree/abianchi/PLAT-5697) - is that intentional? Is there something we'd want to merge first before landing this?

**Review by juj — APPROVED** [review]
LGTM, dropped in some comments, particularly the functionality vs boilerplate ratio is a bit high on the added code; feel free to address whichever comments you find relevant, and then land.

File: add_demo_src/add_demo.py
Comment (rachelmsimm) [reply add_demo_src/add_demo.py:26]: I could see someone wanting to make the key something like `game-build-a` and the name something like `Game Build A` - but I agree that its not necessary. If we are going to have a name it probably shouldn't have to be unique.

I think that change is out of the scope of this PR, though

**Comment by rachelmsimm** [comment]
> I see the base of this PR is to land on top of Andrew Bianchi's feature branch: [abianchi/PLAT-5697](https://github.com/Unity-Technologies/emunittest/tree/abianchi/PLAT-5697) - is that intentional? Is there something we'd want to merge first before landing this?

@juj @mbuscemiunity suggested that I ask you what you prefer. If you think that this would be useful to merge into the main branch of Emunittest then I can change the target branch - otherwise we can just push it to the branch we are using for WPMF

File: add_demo_src/tests/test_add_demo.py
Comment (rachelmsimm) [reply add_demo_src/tests/test_add_demo.py:31]: I think we'd add more tests if we added the actual moving of the demo build into the demos folder in Emunittest, which I think we'd probably want to do at some point. I'd like to keep them here since I think unit tests can be very useful and I don't believe they are obfuscating anything

File: add_demo_src/add_demo.py
Comment (rachelmsimm) [reply add_demo_src/add_demo.py:18]: It is shorter, but I find the separate lines a bit easier to read and debug since `data` has its own variable. I'm not really much of a scripting programmer though 😅

File: .gitignore
Comment (rachelmsimm) [reply .gitignore:10]: I put this in because I was assuming that we wouldn't want to push local updates to the `demo_list.json` file

File: add_demo_src/add_demo.py
Comment (rachelmsimm) [reply add_demo_src/add_demo.py:10]: I'm going to change this to a function instead of an exception 👍

**Comment by rachelmsimm** [comment]
@juj I'm going to push this to the branch that WPMF is using now (abianchi/PLAT-5697) so we can start using it, do you want this merged to main as well?

File: .gitignore
Comment (juj) [reply .gitignore:8]: Oh I see. I wonder if we could then not add the demo_list.json file in the repo at all? Adding a file in the repo, but then simultaneously placing it in .gitignore is something of an untypical mechanism, so would be nice to avoid that kind of pattern.

File: add_demo_src/demo.py
Comment (juj) [reply add_demo_src/demo.py:26]: Either way works - the main thing I was after here is to avoid repetition that needs to list out all the properties in several places, which is a sign of redundant boilerplate in general.

**Comment by juj** [comment]
It would be good to migrate from targeting any work in abianchi's own branches and forks. Instead, let's target all code either in the main branch in this repo, or to a feature branch that mentions WPMF. I suppose you are targeting abianchi/PLAT-5697 since it contains earlier code that are needed for WPMF? If so, then would be good to get those changes merged to either main branch or a branch named wpmf.

File: .gitignore
Comment (rachelmsimm) [reply .gitignore:8]: Gotcha. Would you recommend instead searching for the file in the user's file system and then creating it if it doesn't exist?

**Comment by rachelmsimm** [comment]
> It would be good to migrate from targeting any work in abianchi's own branches and forks. Instead, let's target all code either in the main branch in this repo, or to a feature branch that mentions WPMF. I suppose you are targeting abianchi/PLAT-5697 since it contains earlier code that are needed for WPMF? If so, then would be good to get those changes merged to either main branch or a branch named wpmf.

Yes I have a ticket to rename the branch to be WPMF. Feature branches are for eventually merging into the main branch, right? While I think it could be good to merge this into main, I would defer to your judgment. If you think it makes sense then I'm on board with converting this to a feature branch

File: add_demo_src/demo.py
Comment (rachelmsimm) [reply add_demo_src/demo.py:26]: I respectfully disagree that it is redundant :) Happy to chat more about it if you want though

**Pull request merged**
